tests/fixtures/search.py

```diff
@@ -0,0 +1,293 @@
+            logging.error(
+                "Unable to set up elasticsearch index, search tests will be skipped.",
+                exc_info=e,
+            )
+        return fixture
+
+    def close(self):
+        if self.search:
+            # Delete the works_index, which is almost always created.
+            if self.search.works_index:
+                self.search.indices.delete(self.search.works_index, ignore=[404])
+            # Delete any other indexes created over the course of the test.
+            for index in self.indexes:
+                self.search.indices.delete(index, ignore=[404])
+            ExternalSearchIndex.reset()
+
+    def setup_index(self, new_index):
+        """Create an index and register it to be destroyed during teardown."""
+        self.search.setup_index(new_index=new_index)
+        self.indexes.append(new_index)
+
+    def default_work(self, *args, **kwargs):
+        """Convenience method to create a work with a license pool in the default collection."""
+        work = self.transaction.work(
+            *args,
+            with_license_pool=True,
+            collection=self.transaction.default_collection(),
+            **kwargs
+        )
+        work.set_presentation_ready()
+        return work
+
+
+@pytest.fixture(scope="function")
+def external_search_fixture(
+    database_transaction: DatabaseTransactionFixture,
+) -> Iterable[ExternalSearchFixture]:
+    """Ask for an external search system."""
+    """Note: You probably want EndToEndSearchFixture instead."""
+    data = ExternalSearchFixture.create(database_transaction)
+    yield data
+    data.close()
+
+
+T = TypeVar("T")
+
+
+class EndToEndSearchFixture(Generic[T]):
+    """An external search system fixture that can be populated with data for end-to-end tests."""
+
+    """Tests are expected to call the `populate()` method to populate the fixture with test-specific data."""
+    external_search: ExternalSearchFixture
+    test_data: T
+
+    @classmethod
+    def create(cls, transaction: DatabaseTransactionFixture) -> "EndToEndSearchFixture":
+        data = EndToEndSearchFixture()
+        data.external_search = ExternalSearchFixture.create(transaction)
+        return data
+
+    def populate(self, callback: Callable[["EndToEndSearchFixture[T]"], T]):
+        """Populate the search index with a set of works. The given callback is passed this fixture instance."""
+
+        # Create some works.
+        if not self.external_search.search:
+            # No search index is configured -- nothing to do.
+            return
+
+        self.test_data = callback(self)
+
+        # Add all the works created in the setup to the search index.
+        SearchIndexCoverageProvider(
+            self.external_search.transaction.session(),
+            search_index_client=self.external_search.search,
+        ).run_once_and_update_timestamp()
+
+        # Sleep to give the index time to catch up.
+        time.sleep(2)
```

Comment:
I always wondered why these tests took so long regardless of anything

Reply:
Maybe we can add a speed up
https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-refresh.html

Reply:
We should probably look into whether or not there's any way to have Elasticsearch actually tell us the index is ready.

Reply:
I'm curious if the endpoint blocks until the indices are refreshed.

Reply:
Yes it does, `Refresh requests are synchronous and do not return a response until the refresh operation completes.`

Reply:
I think I'm going to try implementing this, and removing that `sleep` call.

Reply:
This was straightforward to implement, but sadly it made no difference to execution times whatsoever. Presumably it takes exactly as long to `GET` as it does to sleep. Making the `GET` request is preferable though, as it presumably avoids a race condition.

Reply:
We can just use the `indices.refresh()` method.
https://elasticsearch-py.readthedocs.io/en/6.8.2/api.html#elasticsearch.client.IndicesClient.refresh

Reply:
I didn't know about the `refresh()` method! That is really helpful. 

I've traced a flaky test failure in CI to this line before, but I wasn't sure a better way to make sure ES was caught up other then waiting longer 🕐. So I ended up just leaving it. It seems like it would be a win to just call refresh here if it works.

Reply:
I'll switch to `refresh()` instead of that `GET` call.

Reply:
Actually, I'm going to hold off on using `refresh()` as we don't expose that as part of the CM code, and I'm trying to limit changes to the test suite only. It's an easy change to make later.